Use hdf5 file to load catalog data

ahuang314 opened the pull request:
Replaces np.load with h5py.File to open an hdf5 file
Added a dummy weighted catalog file to the repo so that testing can be done. This dummy weighted catalog file contains data for 5 galaxies.

swagnercarena commented on paltax/InputConfigs/input_config.py:
```diff
@@ -103,13 +103,23 @@ def get_config():
-    cosmos_path += '/datasets/cosmos/cosmos_galaxies_train.npz'
+
+    root_path = str(pathlib.Path(__file__).parent.parent.parent)
+    cosmos_path = root_path + '/datasets/weighted_catalog.h5'
```
Please keep this in the cosmos folder

swagnercarena commented on paltax/InputConfigs/input_config.py:
```diff
@@ -103,13 +103,23 @@ def get_config():
+    # These are the weights associated with a cerain parameter
+    # Options are ASYMMETRY, Axial_Ratio, CONCPETRO, GINI, IDENT, M20, RHALFreal, RPETROreal
+    parameter = 'GINI'
+    catalog_weights = hdf5_file['parameters_and_weights'][parameter+'_weights'][:]
```
I think the [:] can be replaced with a .value. Also the formatting is a bit odd here. Why are the images saved in galaxy_images directly whereas the parameters and weights first need to be accessed through 'parameters_and_weights'. I think it would be simpler to:

1. Have all the images, parameters, and weights be keys of the hdf5_file (i.e. directly access gini_weights or gini)
2. Unless there's a very specific reason, I would prefer that all the keys be lowercase. For example, `gini` and `gini_weights` instead of `GINI` as you have it now. Listing all of the options in the CONFIG is a nice touch though.

ahuang314 replied:
It made the most sense to me to group all of the parameters and weights into a common "parameters_and_weights" group, rather than having 16 keys right from the get-go, one for each parameter/weight. But yes, I can change it, these should be easy fixes

swagnercarena commented on paltax/InputConfigs/input_config.py:
```diff
@@ -103,13 +103,23 @@ def get_config():
         'all_subhalo_models': (lens_models.TNFW(),),
         'all_main_deflector_models': (lens_models.EPL(), lens_models.Shear()),
-        'all_source_models': (source_models.CosmosCatalog(cosmos_path),),
+        'all_source_models': (source_models.CosmosCatalog(cosmos_path),
```
We don't want to be using both source models, just the new weightedcatalog model

swagnercarena commented on paltax/InputConfigs/input_config_br.py:
```diff
@@ -101,14 +101,24 @@ def get_config():
         'all_main_deflector_models': (lens_models.EPLEllip(),
                                       lens_models.ShearCart()),
-        'all_source_models': (source_models.CosmosCatalog(cosmos_path),),
+        'all_source_models': (source_models.CosmosCatalog(cosmos_path),
```
See comments on input_config file

swagnercarena commented on paltax/InputConfigs/input_config_br.py:
```diff
@@ -101,14 +101,24 @@ def get_config():
+    # These are the weights associated with a cerain parameter
+    # Options are ASYMMETRY, Axial_Ratio, CONCPETRO, GINI, IDENT, M20, RHALFreal, RPETROreal
+    parameter = 'GINI'
+    catalog_weights = hdf5_file['parameters_and_weights'][parameter+'_weights'][:]
```
See comments on input_config file

swagnercarena commented on paltax/InputConfigs/input_config_psf.py:
```diff
@@ -102,15 +102,25 @@ def get_config():
+    # These are the weights associated with a cerain parameter
+    # Options are ASYMMETRY, Axial_Ratio, CONCPETRO, GINI, IDENT, M20, RHALFreal, RPETROreal
+    parameter = 'GINI'
+    catalog_weights = hdf5_file['parameters_and_weights'][parameter+'_weights'][:]
```
See comments on input_config file

swagnercarena commented on paltax/InputConfigs/input_config_psf.py:
```diff
@@ -102,15 +102,25 @@ def get_config():
                                       lens_models.ShearCart()),
-        'all_source_models': (source_models.CosmosCatalog(cosmos_path),),
+        'all_source_models': (
+            source_models.CosmosCatalog(cosmos_path),
```
See comments on input_config file

swagnercarena commented on paltax/InputConfigs/input_config_test.py:
```diff
@@ -103,13 +103,22 @@ def get_config():
+    # These are the weights associated with a cerain parameter
+    # Options are ASYMMETRY, Axial_Ratio, CONCPETRO, GINI, IDENT, M20, RHALFreal, RPETROreal
+    parameter = 'GINI'
+    catalog_weights = hdf5_file['parameters_and_weights'][parameter+'_weights'][:]
```
See comments on input_config file

swagnercarena commented on paltax/InputConfigs/input_config_test.py:
```diff
@@ -103,13 +103,22 @@ def get_config():
         'all_main_deflector_models': (lens_models.EPL(), lens_models.Shear()),
-        'all_source_models': (source_models.CosmosCatalog(cosmos_path),),
+        'all_source_models': (source_models.CosmosCatalog(cosmos_path),
+                              source_models.WeightedCatalog(cosmos_path, jnp.asarray(catalog_weights))),
```
See comments on input_config file

swagnercarena commented on paltax/source_models.py:
```diff
@@ -23,6 +23,7 @@
 import jax
 import jax.numpy as jnp
 import numpy as np
+import h5py
```
put in alphabetic order

swagnercarena commented on paltax/source_models.py:
```diff
@@ -293,16 +294,20 @@ def modify_cosmology_params(
-        images = npz_file['images']
+
+        hdf5_file = h5py.File(self.cosmos_path, 'r')
+        images = hdf5_file['galaxy_images'][:][:][:]
```
The first [:] we'll probably want to modify to access the current batch of images. After that, I think we want to use .value (avoid the other two [:])

swagnercarena replied:
I leave this up to you, but I think just using the label `images` might be more standard than specifying `galaxy_images`.

swagnercarena commented on paltax/source_models.py:
```diff
@@ -293,16 +294,20 @@ def modify_cosmology_params(
         cosmology_params['cosmos_n_images'] = len(images)
-        redshifts = npz_file['redshifts']
-        pixel_sizes = npz_file['pixel_sizes']
+        redshifts = hdf5_file['redshifts'][:]
```
.value here and below

ahuang314 replied:
Do you mean like this? 
![image](https://github.com/user-attachments/assets/a2986377-d7f2-4681-a3f1-e9609e31859c)
It doesn't work:
![image](https://github.com/user-attachments/assets/e9ca49f4-2aad-492c-b872-f81b25e82aaf)
I also tried .value(), .values, and .values()

swagnercarena replied:
Looks like the .value was deprecated in version 2.1, so you can ignore those comments. Anyways once we implement the ability to load a portion of the dataset the [:] notation would be the appropriate choice.

swagnercarena commented on paltax/InputConfigs/input_config.py:
```diff
@@ -103,13 +103,23 @@ def get_config():
+    root_path = str(pathlib.Path(__file__).parent.parent.parent)
+    cosmos_path = root_path + '/datasets/weighted_catalog.h5'
+    
+    hdf5_file = h5py.File(cosmos_path, 'r')
```
Here an in other init files, use with to open the hdf5 file and automatically close it once the weights are extracted

swagnercarena commented on paltax/source_models.py:
```diff
@@ -293,16 +294,20 @@ def modify_cosmology_params(
-        npz_file = np.load(self.cosmos_path)
-        images = npz_file['images']
+
+        hdf5_file = h5py.File(self.cosmos_path, 'r')
```
Since we're switching to hdf5, I propose three changes.

1. The cosmos catalog should now include a parameter that determines how many images to load when modify_cosmology_params is called. It should have a relatively large default value (say 2048).
2. The hdf5_file should be opened in init and saved as a self property. You should use [atexit](https://docs.python.org/3/library/atexit.html#atexit.register) at the end of the __init__ function to register a cleanup function that will close the hdf5 file once python exits. 
3. You should save a tracker of the number of times modify_cosmology_params has been called and the total length of the hdf5 file. The idea here is that each time we call modify_cosmology_params we should be accessing a new chunk of the hdf5 file. When we run out of chunks, we should start again at the first chunk. Happy to talk on slack if you need more details on this.

The high-level idea here is to make sure that we can use hdf5 files that are larger than the number of images we can realistically store on memory. Note that once you've made that change, we'll also have to change how you currently deal with the weights.

ahuang314 commented:
@swagnercarena OK, I've implemented most of your suggested changes plus some more. Here's what I've done:

1. `WeightedCatalog` no longer takes `catalog_weights` as input during initialization. Now it takes just a paramater (e.g. 'gini'), and the catalog weights are loaded up during the `__init__` function instead. This way we do not have to open the hdf5 file in the input config.
2. Added functionality to load batches of the images at a time. I added a `batch_size` parameter to the `CosmosCatalog` as you suggested. `self.batch_number` keeps track of the number of times `modify_cosmology_params` has been called. `self.total_num_galaxies` keeps track of the size of the hdf5 file so we know when to loop back to the first batch.
3. hdf5 formatting changed as you suggested
4. I have not yet changed any of the test files to accommodate the new structure of the classes. Will work on that once you approve of the new structure

swagnercarena commented on paltax/source_models.py:
```diff
@@ -295,19 +305,25 @@ def modify_cosmology_params(
+        # when loading from hdf5 dataset
+        pixel_sizes = self.hdf5_file['pixel_sizes'][start_val:end_val]
+        redshifts = self.hdf5_file['redshifts'][start_val:end_val]
+        images = self.hdf5_file['galaxy_images'][start_val:end_val][:][:]
```
Also I think you changed the name to images

swagnercarena commented on paltax/source_models.py:
```diff
@@ -442,8 +458,12 @@ def __init__(self, cosmos_path: str, catalog_weights: jnp.ndarray):
+        super().__init__(cosmos_path=cosmos_path, batch_size=batch_size)
+
+        # Loads the catalog weights corresponding to the input parameter
+        catalog_weights = self.hdf5_file[parameter + '_weights'][:]
```
Just save the parameter. Load and process the weights later.

ahuang314 commented on paltax/InputConfigs/input_config.py:
```diff
@@ -105,21 +104,21 @@ def get_config():
 
     root_path = str(pathlib.Path(__file__).parent.parent.parent)
-    cosmos_path = root_path + '/datasets/weighted_catalog.h5'
+    cosmos_path = root_path + '/datasets/cosmos/cosmos_galaxies_train.npz'
```
Ah, gotcha. Yeah, I had gotten confused from the previous comment

ahuang314 commented on paltax/InputConfigs/input_config.py:
```diff
@@ -105,21 +104,21 @@ def get_config():
     config['all_models'] = {
         'all_los_models': (lens_models.NFW(),),
         'all_subhalo_models': (lens_models.TNFW(),),
-        'all_main_deflector_models': (lens_models.EPL(), lens_models.Shear()),
```
Whoops. What happened was I made changes to one of the config files and copy pasted the snippet of code to the other config files, not realizing that they were different.

ahuang314 commented on notebooks/GenerateImages.ipynb:
Yes, I changed this notebook intentionally because it had that cell that was not supposed to be there as you mentioned

swagnercarena commented on paltax/source_models.py:
```diff
@@ -437,8 +461,12 @@ def __init__(self, cosmos_path: str, catalog_weights: jnp.ndarray):
+        super().__init__(cosmos_path=cosmos_path, images_per_chunk=images_per_chunk)
+
+        # Loads the catalog weights corresponding to the input parameter
+        catalog_weights = self.hdf5_file[parameter + '_weights']
```
You should save the parameter value, not the catalog_weights. Technically this would also work, but I think it would be clearer if we accessed the hdf5 file in the modify_cosmology_params call.

swagnercarena commented on paltax/source_models.py:
```diff
@@ -437,8 +461,12 @@ def __init__(self, cosmos_path: str, catalog_weights: jnp.ndarray):
+        catalog_weights = self.hdf5_file[parameter + '_weights']
 
         # Turns the catalog_weights pdf into a normalized cdf
         catalog_weights_cdf = (
```
These needs to be done on the chunks drawn in modify_cosmology_params

swagnercarena reviewed (changes requested):
Looks good, now we just need to modify the tests.

1. Make sure all the old tests are passing. If the failure of a test is confusing to you, we can discuss it on slack / on this thread.
2. We should write a new test to make sure the chunking is well behaved. Essentially we should call modify_cosmology_params enough to loop through the test dataset and confirm the the images we expect to be loaded are the ones that are loaded in each call (i.e. image 0:2, 2:4, 4:5, 0:2, ... if the chunk size is 2 and the dataset has 5 images).

swagnercarena commented on paltax/source_models_test.py:
```diff
@@ -33,10 +34,12 @@
 from paltax import source_models
 from paltax import utils
 
+# NOTE: The tests are done assuming that the test
```
I like the inclusion of this note, but put it closer to where the tests make this assumption instead of the top of the file.

swagnercarena commented on paltax/source_models_test.py:
```diff
@@ -52,12 +55,17 @@
-    catalog_weights = jnp.array([0.2, 10.0])
-    catalog_weights_cdf = jnp.cumsum(catalog_weights) / jnp.sum(catalog_weights)
-    return catalog_weights, catalog_weights_cdf
+def catalog_weights_from_hdf5(parameter):
```
put _ in front of these functions to clarify that they are only for internal use within this test file.

swagnercarena commented on paltax/source_models_test.py:
```diff
@@ -52,12 +55,17 @@
+        catalog_weights = jnp.array(hdf5_file[parameter + '_weights'])
+    return catalog_weights
 
+def compute_catalog_weights_cdf(catalog_weights, start_val, images_per_chunk):
```
See above comment.

swagnercarena commented on paltax/source_models_test.py:
```diff
@@ -298,23 +306,36 @@ class CosmosCatalogTest(chex.TestCase, parameterized.TestCase):
+        self.assertEqual(cosmos_catalog.hdf5_file.filename, COSMOS_TEST_PATH)
+
+        # Test that the correct total number of galaxies is being stored (should be 5)
+        assert cosmos_catalog.total_num_galaxies == 5
```
For consistency, can you use self.assertEqual here and in other places where you use assert

swagnercarena commented on paltax/source_models_test.py:
```diff
@@ -298,23 +306,36 @@ class CosmosCatalogTest(chex.TestCase, parameterized.TestCase):
+        # Test that the correct total number of galaxies is being stored (should be 5)
+        assert cosmos_catalog.total_num_galaxies == 5
+
+        # Test that the WeightedCatalog stores the images per chunk and starts at chunk 0
```
CosmosCatalog not WeightedCatalog

swagnercarena commented on paltax/source_models_test.py:
```diff
@@ -298,23 +306,36 @@ class CosmosCatalogTest(chex.TestCase, parameterized.TestCase):
-        self.assertEqual(cosmology_params['cosmos_n_images'], 2)
+                              (3,))
+        self.assertEqual(cosmology_params['cosmos_n_images'], 3)
+        assert cosmos_catalog.chunk_number == 1
```
see comment on assert

swagnercarena commented on paltax/source_models_test.py:
```diff
@@ -460,67 +481,105 @@ class WeightedCatalogTest(chex.TestCase):
+        self.assertEqual(weighted_catalog.hdf5_file.filename, COSMOS_TEST_PATH)
+
+        # Test that the correct total number of galaxies is being stored (should be 5)
+        assert weighted_catalog.total_num_galaxies == 5
```
see comment on assert

swagnercarena commented on paltax/source_models_test.py:
```diff
@@ -460,67 +481,105 @@ class WeightedCatalogTest(chex.TestCase):
+        np.testing.assert_array_almost_equal(
+            cosmology_params['catalog_weights_cdf'], catalog_weights_cdf
+        )
+        assert weighted_catalog.chunk_number == 1
```
see comment on assert

swagnercarena commented on paltax/source_models_test.py:
```diff
@@ -460,67 +481,105 @@ class WeightedCatalogTest(chex.TestCase):
+            cosmology_params['catalog_weights_cdf'], catalog_weights_cdf
+        )
+        # Since the third chunk is just one number, the cumsum should just be 1
+        assert cosmology_params['catalog_weights_cdf'].size == 1
```
see comment on assert

swagnercarena reviewed (changes requested):
Looks really good. Asked for some small changes but should be good to go after that.

swagnercarena commented:
It looks like the coverage on source_models.py [has dropped](https://coveralls.io/builds/69083919). Would you mind running coverage locally to see what line is missing? 

```
coverage run --source paltax --omit="*Configs/*,*_test.py,main.py" -m unittest discover -v -p "*_test.py"
coverage html
open html/index.rst
```
